chain/x/zoracle/internal/keeper/data_source.go

```diff
@@ -5,10 +5,26 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-// SetDataSource saves the given data source with the given ID to the storage.
-func (k Keeper) SetDataSource(ctx sdk.Context, id int64, dataSource types.DataSource) {
+// AddDataSource adds the given data source to the storage.
+func (k Keeper) AddDataSource(ctx sdk.Context, owner sdk.AccAddress, name string, fee sdk.Coins, executable []byte) sdk.Error {
+	newDataSourceID := k.GetNextDataSourceID(ctx)
+
+	// TODO: check executable size.
+	newDataSource := types.NewDataSource(owner, name, fee, executable)
+
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.DataSourceStoreKey(newDataSourceID), k.cdc.MustMarshalBinaryBare(newDataSource))
+	return nil
+}
+
+// EditDataSource edits the given data source by given data source id to the storage.
+func (k Keeper) EditDataSource(ctx sdk.Context, dataSourceID int64, owner sdk.AccAddress, name string, fee sdk.Coins, executable []byte) sdk.Error {
+	// TODO: check executable size.
+	updatedDataSource := types.NewDataSource(owner, name, fee, executable)
+
 	store := ctx.KVStore(k.storeKey)
-	store.Set(types.DataSourceStoreKey(id), k.cdc.MustMarshalBinaryBare(dataSource))
+	store.Set(types.DataSourceStoreKey(dataSourceID), k.cdc.MustMarshalBinaryBare(updatedDataSource))
+	return nil
```

Comment:
Both function have duplicate code to store dataSource. I think we have `SetDataSource` function to manage key and save to store and 2 functions can call it

Reply:
private?

Reply:
public is ok i think